Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Ride.hs

```diff
@@ -301,6 +302,72 @@ syncCompletedRide ride booking = do
+  bookings <- QBooking.findAllByBookingIds $ map (.bookingId) rides
+  let bookingsRideIdsZip = zip bookings rideIds
+  -- merchant access checking
+  for_ bookingsRideIdsZip $ \(booking, rideId) -> do unless (merchant.id == booking.providerId) $ throwError (RideDoesNotExist rideId.getId)
+  logTagInfo "dashboard -> syncRide : " $ show rideIds <> "; status: " <> show (map fun rides)
+  let ridesBookingsZip = zip rides bookings
+  rideDataResult <-
+    mapM
+      ( \(ride, booking) -> case ride.status of
+          DRide.NEW -> syncNewMultipleRide ride booking
+          DRide.INPROGRESS -> syncInProgressMultipleRide ride booking
+          DRide.COMPLETED -> syncCompletedMultipleRide ride booking
+          DRide.CANCELLED -> syncCancelledMultipleRide ride booking merchant
+      )
+      ridesBookingsZip
+  return
+    Common.MultipleRideSyncRes
+      { list = rideDataResult
+      }
+  where
+    fun ride = ride.status
+
+syncNewMultipleRide :: DRide.Ride -> DBooking.Booking -> Flow Common.MultipleRideData
+syncNewMultipleRide ride booking = do
+  handle (errHandler ride.status booking.status "ride assigned") $
+    CallBAP.sendRideAssignedUpdateToBAP booking ride
+  pure $ Common.MultipleRideData {newStatus = Common.RIDE_NEW, message = "Success. Sent ride started update to bap", rideId = cast @DRide.Ride @Common.Ride ride.id}
+
+syncInProgressMultipleRide :: DRide.Ride -> DBooking.Booking -> Flow Common.MultipleRideData
+syncInProgressMultipleRide ride booking = do
+  handle (errHandler ride.status booking.status "ride started") $
+    CallBAP.sendRideStartedUpdateToBAP booking ride
+  pure $ Common.MultipleRideData {newStatus = Common.RIDE_INPROGRESS, message = "Success. Sent ride started update to bap", rideId = cast @DRide.Ride @Common.Ride ride.id}
+
+syncCancelledMultipleRide :: DRide.Ride -> DBooking.Booking -> DM.Merchant -> Flow Common.MultipleRideData
+syncCancelledMultipleRide ride booking merchant = do
+  mbBookingCReason <- runInReplica $ QBCReason.findByRideId ride.id
+  -- rides cancelled by bap no need to sync, and have mbBookingCReason = Nothing
+  case mbBookingCReason of
+    Nothing -> pure $ Common.MultipleRideData {newStatus = Common.RIDE_CANCELLED, message = "No cancellation reason found for ride. Nothing to sync, ignoring", rideId = cast @DRide.Ride @Common.Ride ride.id}
+    Just bookingCReason -> do
+      case bookingCReason.source of
+        DBCReason.ByUser -> pure $ Common.MultipleRideData {newStatus = Common.RIDE_CANCELLED, message = "Ride canceled by customer. Nothing to sync, ignoring", rideId = cast @DRide.Ride @Common.Ride ride.id}
+        source -> do
+          handle (errHandler ride.status booking.status "booking cancellation") $
+            CallBAP.sendBookingCancelledUpdateToBAP booking merchant source
+          pure $ Common.MultipleRideData {newStatus = Common.RIDE_CANCELLED, message = "Success. Sent booking cancellation update to bap", rideId = cast @DRide.Ride @Common.Ride ride.id}
```

Comment:
These code duplicated a lot with existing functions `syncNewRide`, `syncCancelledRide` and others. But if we going to remove `rideSync` api, then its ok

Reply:
got it